tests/suites/test_suite_x509parse.data

```diff
@@ -2940,6 +2940,18 @@ X509 CSR ASN.1 (OK)
 depends_on:MBEDTLS_PK_CAN_ECDSA_SOME:MBEDTLS_ECP_HAVE_SECP256R1:MBEDTLS_MD_CAN_SHA1:!MBEDTLS_X509_REMOVE_INFO
 mbedtls_x509_csr_parse:"308201183081bf0201003034310b3009060355040613024e4c3111300f060355040a1308506f6c617253534c31123010060355040313096c6f63616c686f73743059301306072a8648ce3d020106082a8648ce3d0301070342000437cc56d976091e5a723ec7592dff206eee7cf9069174d0ad14b5f768225962924ee500d82311ffea2fd2345d5d16bd8a88c26b770d55cd8a2a0efa01c8b4edffa029302706092a864886f70d01090e311a301830090603551d1304023000300b0603551d0f0404030205e0300906072a8648ce3d04010349003046022100b49fd8c8f77abfa871908dfbe684a08a793d0f490a43d86fcf2086e4f24bb0c2022100f829d5ccd3742369299e6294394717c4b723a0f68b44e831b6e6c3bcabf97243":"CSR version   \: 1\nsubject name  \: C=NL, O=PolarSSL, CN=localhost\nsigned using  \: ECDSA with SHA1\nEC key size   \: 256 bits\n\nkey usage         \: Digital Signature, Non Repudiation, Key Encipherment\n":0
 
+X509 CSR ASN.1 (Unsupported critical extension)
+depends_on:MBEDTLS_PK_CAN_ECDSA_SOME:MBEDTLS_ECP_HAVE_SECP256R1:MBEDTLS_MD_CAN_SHA256:!MBEDTLS_X509_REMOVE_INFO
+mbedtls_x509_csr_parse:"308201233081cb02010030413119301706035504030c1053656c66207369676e65642074657374310b300906035504061302444531173015060355040a0c0e41757468437274444220546573743059301306072a8648ce3d020106082a8648ce3d03010703420004c11ebb9951848a436ca2c8a73382f24bbb6c28a92e401d4889b0c361f377b92a8b0497ff2f5a5f6057ae85f704ab1850bef075914f68ed3aeb15a1ff1ebc0dc6a028302606092a864886f70d01090e311930173015060b2b0601040183890c8622020101ff0403010101300a06082a8648ce3d040302034700304402200c4108fd098525993d3fd5b113f0a1ead8750852baf55a2f8e670a22cabc0ba1022034db93a0fcb993912adcf2ea8cb4b66389af30e264d43c0daea03255e45d2ccc":"":MBEDTLS_ERR_X509_INVALID_EXTENSIONS+MBEDTLS_ERR_ASN1_UNEXPECTED_TAG
```

Comment:
I've just tried and if I run this test case on 2.28, it fails.

So the bug is present in 2.28 and we do need a backport, both of this test case and of the backport.

To save a bit of time, in 2.28, the `depends_on` line should be
```
X509 CSR ASN.1 (Unsupported critical extension)
depends_on:MBEDTLS_ECDSA_C:MBEDTLS_ECP_DP_SECP256R1_ENABLED:MBEDTLS_SHA256_C
mbedtls_x509_csr_parse:"308201233081cb02010030413119301706035504030c1053656c66207369676e65642074657374310b300906035504061302444531173015060355040a0c0e41757468437274444220546573743059301306072a8648ce3d020106082a8648ce3d03010703420004c11ebb9951848a436ca2c8a73382f24bbb6c28a92e401d4889b0c361f377b92a8b0497ff2f5a5f6057ae85f704ab1850bef075914f68ed3aeb15a1ff1ebc0dc6a028302606092a864886f70d01090e311930173015060b2b0601040183890c8622020101ff0403010101300a06082a8648ce3d040302034700304402200c4108fd098525993d3fd5b113f0a1ead8750852baf55a2f8e670a22cabc0ba1022034db93a0fcb993912adcf2ea8cb4b66389af30e264d43c0daea03255e45d2ccc":"":MBEDTLS_ERR_X509_INVALID_EXTENSIONS+MBEDTLS_ERR_ASN1_UNEXPECTED_TAG
```

Reply:
Having dug deeper, if I understand correctly, this test case has a **[requested extension](https://datatracker.ietf.org/doc/html/rfc2985#section-5.4)** which is marked as critical. Why should it be rejected? The CA doesn't need to understand requested extensions. That's the job of the verifier. All the CA needs to do is to copy the extensions into the certificate. The `critical` extension request is just a data field, not a policy for the CA.

If my understanding is correct then this test case is wrong: the expected result should be a success, since there's nothing wrong with the requested extensions.

Given the debate here we should have both a test case with a requested extension critical=true and one with critical=false, both expected to succeed. (In addition to the existing test cases with extensions without the `critical` attribute.)

With respect to backporting, we'd still want to backport the test cases to 2.28, and they would pass.

Reply:
Hmm.
I think that the CA needs to understand the extensions to ensure that it's okay to sign the certificate, otherwise you could do the following:

1. Mallory -> CA Bob: "Please sign this CSR with this SubjectAltName extension, `dNSName=importantdomain.com`"
2. CA Bob -> Mallory: [Bob doesn't understand SubjectAltName] "Certainly, here is your certificate with your unknown extension copied in"
3. Mallory has a signed certificate with a domain they don't own, because Bob wasn't aware of the SubjectAltName extension.

This is why we currently ignore all CSR requested extensions that we do not understand, without copying them into the CSR.

Reply:
Hi Gilles,
thanks for your analysis. If you go back to the issue #8377, the original problem was that extensions with the optionally available critical flag were not accepted. Hence, in my initial solution (https://github.com/Mbed-TLS/mbedtls/pull/8378/files/adb3cc4d433466f00ba827d5288577e367fbf1f3#diff-dccd5ffd55ff529be4e531be08342be60a967c90fa9f1ad0aef790e615846148), I justed tried to extract the optional critical flag but then ignored it in case it was available.
In https://github.com/Mbed-TLS/mbedtls/issues/8377#issuecomment-1766389135, Tom argued that we shall not blindly accept critical extensions, in case we do not know them, hence, the callbacks were getting introduced. As no extensions are parsed in mbedTLS < 3.4.0, all extensions critical or not are accepted.

Reply:
Thanks David, that's a very good example. It shows that we must not copy unknown extensions. But fortunately as you write, we don't copy unknown extensions: we skip them. This is true in 2.28 and in 3.5.

With this pull request, we change the behavior of unknown critical extensions to be rejected instead of ignored. With the understanding I now have, this seems reasonable to me.

All of this should be documented, though! This is a preexisting issue, but as is often the case with technical debt, we could get away with it originally, but now that it explodes in our face, we need to resolve it.

So I think what we need to do now is:

* In both 2.28 and here, update `x509_csr.h` to explain how `mbedtls_x509_csr_parse_der` handles requested extensions.
* Have test cases, in both 2.28 and here, with:
    * An unknown extension without the critical attribute.
    * An unknown extension with critical=false.
    * An unknown extension with critical=true.

Reply:
Hi Gilles,

I now updated the mbedTLS 3 test cases:
- An unknown extension without the critical attribute.
  - X509 CSR ASN.1 (Unsupported non-critical extension, critical undefined)
  - [Test data](http://lapo.it/asn1js/#MIIBIjCByAIBADBBMRkwFwYDVQQDDBBTZWxmIHNpZ25lZCB0ZXN0MQswCQYDVQQGEwJERTEXMBUGA1UECgwOQXV0aENydERCIFRlc3QwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAARflLKNEzQYgzvxDEQtkTBkWdOSXnzqBuu5Igky594Rb7ZxxdLWwKN4ShKJcheu-EMucij86gqwFr22e2fO1MYSoCUwIwYJKoZIhvcNAQkOMRYwFDASBgsrBgEEAYOJDIYiAgQDAQEBMAoGCCqGSM49BAMCA0kAMEYCIQCbHosld1wYUl6WdT4e1Vh1-NYvAmxbf3DrUDetJ9yS3gIhALod_hTeavamA_djVj_QRrHNNxS1TW2vXYpyB2SX8RAU)
  - ![grafik](https://github.com/Mbed-TLS/mbedtls/assets/140500342/e94d0d09-9b93-431d-bcd2-89614d0ab6e6)
- An unknown extension with critical=false.
  - X509 CSR ASN.1 (Unsupported non-critical extension, critical=false)
  - [Test data](http://lapo.it/asn1js/#MIIBJDCBywIBADBBMRkwFwYDVQQDDBBTZWxmIHNpZ25lZCB0ZXN0MQswCQYDVQQGEwJERTEXMBUGA1UECgwOQXV0aENydERCIFRlc3QwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAAS15xin32shkScz53xC8ma4KD5srnrfiv1WuZDBxiMuoKKkYJfCGDU7yUhESuo9AEI-hIAuKMSAmWQf6Zd837UFoCgwJgYJKoZIhvcNAQkOMRkwFzAVBgsrBgEEAYOJDIYiAgEBAAQDAQEBMAoGCCqGSM49BAMCA0gAMEUCIQDwupoIRq0KfO_Qph1fySGU3AYDekQVjeLQxWmRLAWNQwIgQh8nqfJJwWh-L6NNs_VDyFEv2SXf5a4Ahn8Tlj_9T40)
  - ![grafik](https://github.com/Mbed-TLS/mbedtls/assets/140500342/782a19e4-fdd6-4f9a-b2c6-563578c56d67)
  - Test CSR was generated using mbedTLS after editing x509_write_extension to always add the optional critical field:
  ```
    if (ext->val.p[0] != 0) {
        MBEDTLS_ASN1_CHK_ADD(len, mbedtls_asn1_write_bool(p, start, 1));
    } else {
        MBEDTLS_ASN1_CHK_ADD(len, mbedtls_asn1_write_bool(p, start, 0));
    }
  ```
- An unknown extension with critical=true.
  - X509 CSR ASN.1 (Unsupported critical extension, critical=true)
  - X509 CSR ASN.1 (Unsupported critical extension accepted by callback, critical=true)
  - X509 CSR ASN.1 (Unsupported critical extension rejected by callback, critical=true)
  - [Test data](http://lapo.it/asn1js/#MIIBIzCBywIBADBBMRkwFwYDVQQDDBBTZWxmIHNpZ25lZCB0ZXN0MQswCQYDVQQGEwJERTEXMBUGA1UECgwOQXV0aENydERCIFRlc3QwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAATBHruZUYSKQ2yiyKczgvJLu2woqS5AHUiJsMNh83e5KosEl_8vWl9gV66F9wSrGFC-8HWRT2jtOusVof8evA3GoCgwJgYJKoZIhvcNAQkOMRkwFzAVBgsrBgEEAYOJDIYiAgEB_wQDAQEBMAoGCCqGSM49BAMCA0cAMEQCIAxBCP0JhSWZPT_VsRPwoerYdQhSuvVaL45nCiLKvAuhAiA025Og_LmTkSrc8uqMtLZjia8w4mTUPA2uoDJV5F0szA)
  - ![grafik](https://github.com/Mbed-TLS/mbedtls/assets/140500342/3c8bdcc3-f140-4e07-bdfc-5880846f1d29)

Reply:
In case the behavior whall be described in x509_csr.h, then the behavior or mbedtls_x509_crt_parse_der should also be described in x509_crt.h.

Reply:
Re documenting this, I think in 2.28 the description of `mbedtls_x509_csr_parse_der()` is already correct, as [it says](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-2.28/include/mbedtls/x509_csr.h#L71):
```
* \note           CSR attributes (if any) are currently silently ignored.
```
Although perhaps it could be made more clear that this includes the requested extensions. However, the testcases would still need adding so this doesn't save us anything.

Reply:
In the development branch the same note exists for this function. So shall we simply remove it?
https://github.com/Mbed-TLS/mbedtls/blob/development/include/mbedtls/x509_csr.h#L78

Reply:
In `development` we actually have changed the behaviour here so should change it to something like:
```
* \note

Reply:
Backport PR (https://github.com/Mbed-TLS/mbedtls/pull/8491) for the tests is ready. However, tests currently fail. This is expected as mbedtls < 3.4.0 does not care whether unsupported critical extensions exist of not when parsing a CSR.

Reply:
> In `development` we actually have changed the behaviour here so should change it to something like:
> 
> ```
> * \note          Any unsupported requested extensions are ignored, unless the critical flag
> *                is set, in which case the CSR is rejected.
> ```

Done.